packages/roosterjs-editor-core/lib/undo/Undo.ts

```diff
@@ -0,0 +1,226 @@
+            case PluginEventType.ContentChanged:
+                if ((<ContentChangedEvent>event).source != 'Undo' && !this.isRestoring) {
+                    this.clearRedoForInput();
+                }
+                break;
+        }
+    }
+
+    /**
+     * Clear all existing undo snapshots
+     */
+    public clear() {
+        this.undoSnapshots = null;
+        this.hasNewContent = false;
+    }
+
+    /**
+     * Restore an undo snapshot to editor
+     */
+    public undo(): void {
+        if (this.hasNewContent) {
+            this.addUndoSnapshot();
+        }
+
+        this.restoreSnapshot(-1 /*previousSnapshot*/);
+    }
+
+    /**
+     * Restore a redo snapshot to editor
+     */
+    public redo(): void {
+        this.restoreSnapshot(1 /*nextSnapshot*/);
+    }
+
+    /**
+     * Whether there is a snapshot for undo
+     */
+    public canUndo(): boolean {
+        return this.hasNewContent || this.getSnapshotsManager().canMove(-1 /*previousSnapshot*/);
+    }
+
+    /**
+     * Whether there is a snapshot for redo
+     */
+    public canRedo(): boolean {
+        return this.getSnapshotsManager().canMove(1 /*nextSnapshot*/);
+    }
+
+    /**
+     * Add an undo snapshot
+     */
+    public addUndoSnapshot(): void {
+        let snapshot = buildSnapshot(this.editor);
+        this.getSnapshotsManager().addSnapshot(snapshot);
+        this.hasNewContent = false;
+    }
+
+    private restoreSnapshot(delta: number) {
+        let snapshot = this.getSnapshotsManager().move(delta);
+
+        if (snapshot) {
+            try {
+                this.isRestoring = true;
+                restoreSnapshot(this.editor, snapshot);
+            } finally {
+                this.isRestoring = false;
+            }
+
+            this.fireContentChangedEvent();
+        }
+    }
+
+    private onKeyDown(pluginEvent: PluginDomEvent): void {
+        // Handle backspace/delete when there is a selection to take a snapshot
+        // since we want the state prior to deletion restorable
+        let evt = pluginEvent.rawEvent as KeyboardEvent;
+        let shouldTakeUndo = false;
+        if (evt.which == KEY_BACKSPACE || evt.which == KEY_DELETE) {
+            let selectionRange = this.editor.getSelectionRange();
+            if (
+                selectionRange &&
+                (!selectionRange.collapsed ||
+                    // If the selection contains image, we need to add undo snapshots
+                    containsImage(selectionRange.startContainer))
+            ) {
+                shouldTakeUndo = true;
+            }
+        } else if (this.hasNewContent && evt.which >= KEY_PAGEUP && evt.which <= KEY_DOWN) {
+            // PageUp, PageDown, Home, End, Left, Right, Up, Down
+            shouldTakeUndo = true;
+        }
+
+        if (shouldTakeUndo) {
+            this.addUndoSnapshot();
+            this.lastKeyPress = 0;
+        }
+    }
+
+    private onKeyPress(pluginEvent: PluginDomEvent): void {
```

Comment:
What's the difference between keyDown and keyPress event? Why does some keys be handled in keyDown event, but some keys be handled in keyPress event?

Reply:
keyDown -> Key "bottoms out". Will repeat infinitely if the key is held down, per OS behaviour

 keyPress -> a combination of the key going down and then coming up again. It's a single discrete action. There might be many keyDown events, but only one keyPress.

Reply:
There purpose here is to differentiate Input key and Function key. Such as Up/Down/Left/Right/..., they will not trigger keyPress but will trigger keyDown only. Here for KeyPress I will only handle a real input.